File: java/omnia/data/structure/immutable/AbstractBuilder.java
package omnia.data.structure.immutable;

import omnia.data.structure.mutable.LinkedList;
import omnia.data.structure.mutable.MutableList;

abstract class AbstractBuilder<E, B extends AbstractBuilder<E, B, R>, R> {
  final MutableList<E> elements = new LinkedList<>();

  public B add(E element) {
    elements.add(element);
    return getSelf();
  }

  public B addAll(Iterable<? extends E> elements) {
    for (E element : elements) {
      this.elements.add(element);
    }
    return getSelf();
  }

  @SafeVarargs
  public final B addAll(E... elements) {
    for (E element : elements) {
      this.elements.add(element);
    }
    return getSelf();
  }

  public abstract R build();

  protected abstract B getSelf();
}

File: src/omnia/data/structure/immutable/AbstractBuilder.kt
package omnia.data.structure.immutable

import omnia.data.structure.mutable.LinkedList
import omnia.data.structure.mutable.MutableList

abstract class AbstractBuilder<E, B : AbstractBuilder<E, B, R>, R> {
  val elements: MutableList<E> = LinkedList()
  fun add(element: E): B {
    elements.add(element)
    return self
  }

  fun addAll(elements: Iterable<E>): B {
    for (element in elements) {
      this.elements.add(element)
    }
    return self
  }

  @SafeVarargs
  fun addAll(vararg elements: E): B {
    for (element in elements) {
      this.elements.add(element)
    }
    return self
  }

  abstract fun build(): R
  protected abstract val self: B
}